File: contracts/solana/universal_nft/shared/MessageBridge.ts
import { PublicKey } from '@solana/web3.js';
import { ethers } from 'ethers';
import { CrossChainMessageUtils, NFTTransferMessage, CHAIN_IDS } from './CrossChainMessage';

/**
 * Bridge utility for converting between different message formats
 * This module handles conversion between the shared CrossChainMessage format
 * and various chain-specific formats (Solana, EVM, etc.)
 */

/**
 * Solana-specific message format (matches the Rust program structure)
 */
export interface SolanaNftMessage {
  messageType: number;           // u8
  tokenId: string;              // u64 as string
  metadataUri: string;          // String
  name: string;                 // String  
  symbol: string;               // String
  originChainId: number;        // u8
  originAddress: Uint8Array;    // [u8; 32]
  recipientAddress: Uint8Array; // [u8; 32]
  timestamp: number;            // i64
  messageId: Uint8Array;        // [u8; 32]
  additionalMetadata?: Uint8Array; // Option<Vec<u8>>
}

/**
 * Message Bridge class for format conversions
 */
export class MessageBridge {
  
  /**
   * Convert shared NFTTransferMessage to Solana's internal format
   */
  static toSolanaFormat(message: NFTTransferMessage): SolanaNftMessage {
    return {
      messageType: 1, // NFT_MINT
      tokenId: message.tokenId,
      metadataUri: message.metadataUri,
      name: this.extractNameFromUri(message.metadataUri),
      symbol: "UNFT", // Default symbol
      originChainId: message.originChain,
      originAddress: message.originContract,
      recipientAddress: message.recipient,
      timestamp: message.timestamp,
      messageId: message.messageId,
      additionalMetadata: undefined,
    };
  }

  /**
   * Convert Solana format to shared NFTTransferMessage
   */
  static fromSolanaFormat(message: SolanaNftMessage): NFTTransferMessage {
    return {
      tokenId: message.tokenId,
      metadataUri: message.metadataUri,
      recipient: message.recipientAddress,
      originChain: message.originChainId,
      destinationChain: 0, // Will be set by caller
      messageId: message.messageId,
      timestamp: message.timestamp,
      originContract: message.originAddress,
      nonce: "0", // Will be set by caller
    };
  }

  /**
   * Convert shared NFTTransferMessage to EVM ABI format
   * This creates properly ABI-encoded data for Solidity contracts
   */
  static toEvmFormat(message: NFTTransferMessage): string {
    return ethers.AbiCoder.defaultAbiCoder().encode(
      [
        'uint256',  // tokenId
        'string',   // metadataUri
        'bytes32',  // recipient
        'uint32',   // originChain
        'uint32',   // destinationChain
        'bytes32',  // messageId
        'uint64',   // timestamp
        'bytes32',  // originContract
        'uint256',  // nonce
      ],
      [
        message.tokenId,
        message.metadataUri,
        message.recipient,
        message.originChain,
        message.destinationChain,
        message.messageId,
        message.timestamp,
        message.originContract,
        message.nonce,
      ]
    );
  }

  /**
   * Convert EVM ABI format to shared NFTTransferMessage
   */
  static fromEvmFormat(encodedData: string): NFTTransferMessage {
    const decoded = ethers.AbiCoder.defaultAbiCoder().decode(
      [
        'uint256',  // tokenId
        'string',   // metadataUri
        'bytes32',  // recipient
        'uint32',   // originChain
        'uint32',   // destinationChain
        'bytes32',  // messageId
        'uint64',   // timestamp
        'bytes32',  // originContract
        'uint256',  // nonce
      ],
      encodedData
    );

    return {
      tokenId: decoded[0].toString(),
      metadataUri: decoded[1],
      recipient: ethers.getBytes(decoded[2]),
      originChain: decoded[3],
      destinationChain: decoded[4],
      messageId: ethers.getBytes(decoded[5]),
      timestamp: Number(decoded[6]),
      originContract: ethers.getBytes(decoded[7]),
      nonce: decoded[8].toString(),
    };
  }

  /**
   * Create a cross-chain message for NFT transfer from Solana
   */
  static createSolanaToEvmMessage(params: {
    tokenId: string;
    metadataUri: string;
    recipientAddress: string; // EVM address
    destinationChain: number;
    senderAddress: string; // Solana address
    nonce: string;
  }): NFTTransferMessage {
    const timestamp = Math.floor(Date.now() / 1000);
    const senderPubkey = new PublicKey(params.senderAddress);
    const recipientBytes32 = CrossChainMessageUtils.ethereumAddressToBytes32(params.recipientAddress);
    const originContractBytes32 = CrossChainMessageUtils.solanaPublicKeyToBytes32(senderPubkey);

    const messageId = CrossChainMessageUtils.generateMessageId(
      params.senderAddress,
      params.tokenId,
      params.destinationChain,
      params.nonce,
      timestamp
    );

    return {
      tokenId: params.tokenId,
      metadataUri: params.metadataUri,
      recipient: recipientBytes32,
      originChain: CHAIN_IDS.SOLANA_DEVNET,
      destinationChain: params.destinationChain,
      messageId,
      timestamp,
      originContract: originContractBytes32,
      nonce: params.nonce,
    };
  }

  /**
   * Create a cross-chain message for NFT transfer from EVM to Solana
   */
  static createEvmToSolanaMessage(params: {
    tokenId: string;
    metadataUri: string;
    recipientAddress: string; // Solana address
    originChain: number;
    senderAddress: string; // EVM address
    originContractAddress: string; // EVM contract address
    nonce: string;
  }): NFTTransferMessage {
    const timestamp = Math.floor(Date.now() / 1000);
    const recipientPubkey = new PublicKey(params.recipientAddress);
    const recipientBytes32 = CrossChainMessageUtils.solanaPublicKeyToBytes32(recipientPubkey);
    const originContractBytes32 = CrossChainMessageUtils.ethereumAddressToBytes32(params.originContractAddress);

    const messageId = CrossChainMessageUtils.generateMessageId(
      params.senderAddress,
      params.tokenId,
      CHAIN_IDS.SOLANA_DEVNET,
      params.nonce,
      timestamp
    );

    return {
      tokenId: params.tokenId,
      metadataUri: params.metadataUri,
      recipient: recipientBytes32,
      originChain: params.originChain,
      destinationChain: CHAIN_IDS.SOLANA_DEVNET,
      messageId,
      timestamp,
      originContract: originContractBytes32,
      nonce: params.nonce,
    };
  }

  /**
   * Create a cross-chain message for NFT transfer between EVM chains
   */
  static createEvmToEvmMessage(params: {
    tokenId: string;
    metadataUri: string;
    recipientAddress: string; // EVM address
    originChain: number;
    destinationChain: number;
    senderAddress: string; // EVM address
    originContractAddress: string; // EVM contract address
    nonce: string;
  }): NFTTransferMessage {
    const timestamp = Math.floor(Date.now() / 1000);
    const recipientBytes32 = CrossChainMessageUtils.ethereumAddressToBytes32(params.recipientAddress);
    const originContractBytes32 = CrossChainMessageUtils.ethereumAddressToBytes32(params.originContractAddress);

    const messageId = CrossChainMessageUtils.generateMessageId(
      params.senderAddress,
      params.tokenId,
      params.destinationChain,
      params.nonce,
      timestamp
    );

    return {
      tokenId: params.tokenId,
      metadataUri: params.metadataUri,
      recipient: recipientBytes32,
      originChain: params.originChain,
      destinationChain: params.destinationChain,
      messageId,
      timestamp,
      originContract: originContractBytes32,
      nonce: params.nonce,
    };
  }

  /**
   * Validate cross-chain message format compatibility
   */
  static validateCrossChainCompatibility(message: NFTTransferMessage): void {
    // Check token ID is valid
    const tokenIdBN = BigInt(message.tokenId);
    if (tokenIdBN === 0n) {
      throw new Error('Invalid token ID: cannot be zero');
    }

    // Check metadata URI is not empty and not too long
    if (!message.metadataUri || message.metadataUri.length === 0) {
      throw new Error('Metadata URI cannot be empty');
    }
    if (message.metadataUri.length > 500) {
      throw new Error('Metadata URI too long (max 500 characters)');
    }

    // Check timestamp is reasonable (not too old, not too far in future)
    const currentTime = Math.floor(Date.now() / 1000);
    const timeDiff = Math.abs(message.timestamp - currentTime);
    if (timeDiff > 86400) { // 24 hours
      throw new Error('Message timestamp too old or too far in future');
    }

    // Check origin and destination chains are different
    if (message.originChain === message.destinationChain) {
      throw new Error('Origin and destination chains cannot be the same');
    }

    // Check message ID is not all zeros
    const allZeros = new Uint8Array(32);
    if (Buffer.from(message.messageId).equals(Buffer.from(allZeros))) {
      throw new Error('Message ID cannot be all zeros');
    }

    // Check addresses are properly formatted
    if (message.recipient.length !== 32) {
      throw new Error('Recipient address must be 32 bytes');
    }
    if (message.originContract.length !== 32) {
      throw new Error('Origin contract address must be 32 bytes');
    }
  }

  /**
   * Extract NFT name from metadata URI (simplified implementation)
   * In production, you might want to fetch the metadata and extract the name
   */
  private static extractNameFromUri(uri: string): string {
    // This is a simplified implementation
    // You could enhance this to actually fetch the metadata JSON
    const fileName = uri.split('/').pop() || 'Universal NFT';
    return fileName.replace(/\.[^/.]+$/, ''); // Remove file extension
  }

  /**
   * Get supported transfer routes
   */
  static getSupportedRoutes(): Array<{from: number, to: number, name: string}> {
    return [
      { from: CHAIN_IDS.ZETACHAIN_TESTNET, to: CHAIN_IDS.BASE_SEPOLIA, name: 'ZetaChain → Base Sepolia' },
      { from: CHAIN_IDS.ZETACHAIN_TESTNET, to: CHAIN_IDS.SOLANA_DEVNET, name: 'ZetaChain → Solana Devnet' },
      { from: CHAIN_IDS.BASE_SEPOLIA, to: CHAIN_IDS.ZETACHAIN_TESTNET, name: 'Base Sepolia → ZetaChain' },
      { from: CHAIN_IDS.BASE_SEPOLIA, to: CHAIN_IDS.SOLANA_DEVNET, name: 'Base Sepolia → Solana Devnet' },
      { from: CHAIN_IDS.SOLANA_DEVNET, to: CHAIN_IDS.ZETACHAIN_TESTNET, name: 'Solana Devnet → ZetaChain' },
      { from: CHAIN_IDS.SOLANA_DEVNET, to: CHAIN_IDS.BASE_SEPOLIA, name: 'Solana Devnet → Base Sepolia' },
    ];
  }

  /**
   * Check if a transfer route is supported
   */
  static isRouteSupported(fromChain: number, toChain: number): boolean {
    return this.getSupportedRoutes().some(route => 
      route.from === fromChain && route.to === toChain
    );
  }

  /**
   * Estimate gas costs for cross-chain transfer (placeholder implementation)
   */
  static async estimateGasCosts(params: {
    fromChain: number;
    toChain: number;
    provider?: ethers.Provider;
  }): Promise<{
    estimatedGas: string;
    estimatedCostUSD: string;
  }> {
    // This is a placeholder implementation
    // In production, you would integrate with actual gas estimation APIs
    
    const baseGas: Record<number, string> = {
      [CHAIN_IDS.ZETACHAIN_TESTNET]: '100000',
      [CHAIN_IDS.BASE_SEPOLIA]: '150000', 
      [CHAIN_IDS.SOLANA_DEVNET]: '5000', // Different units for Solana
    };

    const estimatedGas = baseGas[params.toChain] || '100000';
    const estimatedCostUSD = '0.50'; // Placeholder

    return {
      estimatedGas,
      estimatedCostUSD,
    };
  }
}

/**
 * Helper functions for testing and development
 */
export class MessageBridgeTestUtils {
  
  /**
   * Create a mock NFT transfer message for testing
   */
  static createMockMessage(overrides: Partial<NFTTransferMessage> = {}): NFTTransferMessage {
    const defaultMessage: NFTTransferMessage = {
      tokenId: "123",
      metadataUri: "https://example.com/metadata/123.json",
      recipient: new Uint8Array(32).fill(1),
      originChain: CHAIN_IDS.ZETACHAIN_TESTNET,
      destinationChain: CHAIN_IDS.BASE_SEPOLIA,
      messageId: new Uint8Array(32).fill(2),
      timestamp: Math.floor(Date.now() / 1000),
      originContract: new Uint8Array(32).fill(3),
      nonce: "1",
    };

    return { ...defaultMessage, ...overrides };
  }

  /**
   * Generate test vectors for cross-chain compatibility testing
   */
  static generateTestVectors(): Array<{
    name: string;
    message: NFTTransferMessage;
    expectedSolanaFormat: SolanaNftMessage;
    expectedEvmFormat: string;
  }> {
    const testMessage = this.createMockMessage();
    
    return [{
      name: "Basic NFT Transfer",
      message: testMessage,
      expectedSolanaFormat: MessageBridge.toSolanaFormat(testMessage),
      expectedEvmFormat: MessageBridge.toEvmFormat(testMessage),
    }];
  }

  /**
   * Validate round-trip conversion consistency
   */
  static validateRoundTrip(originalMessage: NFTTransferMessage): {
    isValid: boolean;
    errors: string[];
  } {
    const errors: string[] = [];

    try {
      // Test EVM round-trip
      const evmEncoded = MessageBridge.toEvmFormat(originalMessage);
      const evmDecoded = MessageBridge.fromEvmFormat(evmEncoded);
      
      if (originalMessage.tokenId !== evmDecoded.tokenId) {
        errors.push('EVM round-trip failed: tokenId mismatch');
      }
      if (originalMessage.metadataUri !== evmDecoded.metadataUri) {
        errors.push('EVM round-trip failed: metadataUri mismatch');
      }
      // Add more comparisons as needed...

      // Test Solana round-trip
      const solanaFormat = MessageBridge.toSolanaFormat(originalMessage);
      const solanaDecoded = MessageBridge.fromSolanaFormat(solanaFormat);
      
      if (originalMessage.tokenId !== solanaDecoded.tokenId) {
        errors.push('Solana round-trip failed: tokenId mismatch');
      }
      // Add more comparisons as needed...

    } catch (error) {
      errors.push(`Round-trip validation error: ${error instanceof Error ? error.message : String(error)}`);
    }

    return {
      isValid: errors.length === 0,
      errors,
    };
  }
}

File: shared/MessageBridge.rs
use anchor_lang::prelude::*;
use solana_program::keccak;
use crate::CrossChainMessage::NFTTransferMessage;

/// Bridge utility for converting between different message formats
/// This module handles conversion between the shared CrossChainMessage format
/// and various chain-specific formats (Solana, EVM, etc.)

/// Convert shared NFTTransferMessage to Solana's internal format
pub fn to_solana_format(message: &NFTTransferMessage) -> SolanaNftMessage {
    SolanaNftMessage {
        message_type: 1, // NFT_MINT
        token_id: message.token_id,
        metadata_uri: message.metadata_uri.clone(),
        name: extract_name_from_uri(&message.metadata_uri), // Extract from metadata if available
        symbol: "UNFT".to_string(), // Default symbol
        origin_chain_id: message.origin_chain as u8,
        origin_address: message.origin_contract,
        recipient_address: message.recipient,
        timestamp: message.timestamp as i64,
        message_id: message.message_id,
        additional_metadata: None,
    }
}

/// Convert Solana format to shared NFTTransferMessage
pub fn from_solana_format(message: &SolanaNftMessage) -> Result<NFTTransferMessage> {
    Ok(NFTTransferMessage {
        token_id: message.token_id,
        metadata_uri: message.metadata_uri.clone(),
        recipient: message.recipient_address,
        origin_chain: message.origin_chain_id as u32,
        destination_chain: 0, // Will be set by caller
        message_id: message.message_id,
        timestamp: message.timestamp as u64,
        origin_contract: message.origin_address,
        nonce: 0, // Will be set by caller
    })
}

/// Convert shared NFTTransferMessage to EVM ABI format
pub fn to_evm_format(message: &NFTTransferMessage) -> Vec<u8> {
    // Using ethers-like ABI encoding
    let mut encoded = Vec::new();
    
    // Encode all fields according to Solidity ABI
    // tokenId (uint256 - 32 bytes)
    let mut token_id_bytes = [0u8; 32];
    token_id_bytes[24..].copy_from_slice(&message.token_id.to_be_bytes());
    encoded.extend_from_slice(&token_id_bytes);
    
    // Calculate dynamic offset for string data
    let base_offset = 32 * 9; // 9 fixed-size fields * 32 bytes each
    let mut dynamic_offset = base_offset;
    
    // metadataUri offset (uint256 - 32 bytes)
    let mut uri_offset_bytes = [0u8; 32];
    uri_offset_bytes[28..].copy_from_slice(&(dynamic_offset as u32).to_be_bytes());
    encoded.extend_from_slice(&uri_offset_bytes);
    
    // Update offset for next dynamic field
    let uri_length = message.metadata_uri.len();
    let uri_padded_length = ((uri_length + 31) / 32) * 32; // Round up to 32-byte boundary
    dynamic_offset += 32 + uri_padded_length; // 32 for length + padded data
    
    // recipient (bytes32)
    encoded.extend_from_slice(&message.recipient);
    
    // originChain (uint32 - padded to 32 bytes)
    let mut origin_chain_bytes = [0u8; 32];
    origin_chain_bytes[28..].copy_from_slice(&message.origin_chain.to_be_bytes());
    encoded.extend_from_slice(&origin_chain_bytes);
    
    // destinationChain (uint32 - padded to 32 bytes)
    let mut dest_chain_bytes = [0u8; 32];
    dest_chain_bytes[28..].copy_from_slice(&message.destination_chain.to_be_bytes());
    encoded.extend_from_slice(&dest_chain_bytes);
    
    // messageId (bytes32)
    encoded.extend_from_slice(&message.message_id);
    
    // timestamp (uint64 - padded to 32 bytes)
    let mut timestamp_bytes = [0u8; 32];
    timestamp_bytes[24..].copy_from_slice(&message.timestamp.to_be_bytes());
    encoded.extend_from_slice(&timestamp_bytes);
    
    // originContract (bytes32)
    encoded.extend_from_slice(&message.origin_contract);
    
    // nonce (uint256 - 32 bytes)
    let mut nonce_bytes = [0u8; 32];
    nonce_bytes[24..].copy_from_slice(&message.nonce.to_be_bytes());
    encoded.extend_from_slice(&nonce_bytes);
    
    // Now encode the dynamic data (metadataUri)
    // Length of string (uint256)
    let mut uri_length_bytes = [0u8; 32];
    uri_length_bytes[28..].copy_from_slice(&(uri_length as u32).to_be_bytes());
    encoded.extend_from_slice(&uri_length_bytes);
    
    // String data (padded to 32-byte boundary)
    let mut uri_data = message.metadata_uri.as_bytes().to_vec();
    while uri_data.len() % 32 != 0 {
        uri_data.push(0); // Pad with zeros
    }
    encoded.extend_from_slice(&uri_data);
    
    encoded
}

/// Convert EVM ABI format to shared NFTTransferMessage
pub fn from_evm_format(data: &[u8]) -> Result<NFTTransferMessage> {
    if data.len() < 32 * 9 {
        return Err(ProgramError::InvalidInstructionData.into());
    }
    
    let mut offset = 0;
    
    // tokenId (uint256)
    let token_id = u64::from_be_bytes(data[offset + 24..offset + 32].try_into().unwrap());
    offset += 32;
    
    // metadataUri offset (uint256) - we'll read the actual string later
    let uri_offset = u32::from_be_bytes(data[offset + 28..offset + 32].try_into().unwrap()) as usize;
    offset += 32;
    
    // recipient (bytes32)
    let recipient = data[offset..offset + 32].try_into().unwrap();
    offset += 32;
    
    // originChain (uint32)
    let origin_chain = u32::from_be_bytes(data[offset + 28..offset + 32].try_into().unwrap());
    offset += 32;
    
    // destinationChain (uint32)
    let destination_chain = u32::from_be_bytes(data[offset + 28..offset + 32].try_into().unwrap());
    offset += 32;
    
    // messageId (bytes32)
    let message_id = data[offset..offset + 32].try_into().unwrap();
    offset += 32;
    
    // timestamp (uint64)
    let timestamp = u64::from_be_bytes(data[offset + 24..offset + 32].try_into().unwrap());
    offset += 32;
    
    // originContract (bytes32)
    let origin_contract = data[offset..offset + 32].try_into().unwrap();
    offset += 32;
    
    // nonce (uint256)
    let nonce = u64::from_be_bytes(data[offset + 24..offset + 32].try_into().unwrap());
    offset += 32;
    
    // Read metadataUri from dynamic data
    if uri_offset >= data.len() {
        return Err(ProgramError::InvalidInstructionData.into());
    }
    
    let uri_length = u32::from_be_bytes(data[uri_offset + 28..uri_offset + 32].try_into().unwrap()) as usize;
    if uri_offset + 32 + uri_length > data.len() {
        return Err(ProgramError::InvalidInstructionData.into());
    }
    
    let metadata_uri = String::from_utf8(
        data[uri_offset + 32..uri_offset + 32 + uri_length].to_vec()
    ).map_err(|_| ProgramError::InvalidInstructionData)?;
    
    Ok(NFTTransferMessage {
        token_id,
        metadata_uri,
        recipient,
        origin_chain,
        destination_chain,
        message_id,
        timestamp,
        origin_contract,
        nonce,
    })
}

/// Solana-specific message format (matches existing program structure)
#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug, PartialEq)]
pub struct SolanaNftMessage {
    pub message_type: u8,
    pub token_id: u64,
    pub metadata_uri: String,
    pub name: String,
    pub symbol: String,
    pub origin_chain_id: u8,
    pub origin_address: [u8; 32],
    pub recipient_address: [u8; 32],
    pub timestamp: i64,
    pub message_id: [u8; 32],
    pub additional_metadata: Option<Vec<u8>>,
}

/// Validate cross-chain message format compatibility
pub fn validate_cross_chain_compatibility(message: &NFTTransferMessage) -> Result<()> {
    // Check token ID is valid
    if message.token_id == 0 {
        return Err(ProgramError::InvalidInstructionData.into());
    }
    
    // Check metadata URI is not empty and not too long
    if message.metadata_uri.is_empty() || message.metadata_uri.len() > 500 {
        return Err(ProgramError::InvalidInstructionData.into());
    }
    
    // Check timestamp is reasonable (not too old, not too far in future)
    let current_time = Clock::get()?.unix_timestamp as u64;
    let time_diff = if message.timestamp > current_time {
        message.timestamp - current_time
    } else {
        current_time - message.timestamp
    };
    
    if time_diff > 86400 { // 24 hours
        return Err(ProgramError::InvalidInstructionData.into());
    }
    
    // Check origin and destination chains are different
    if message.origin_chain == message.destination_chain {
        return Err(ProgramError::InvalidInstructionData.into());
    }
    
    // Check message ID is not all zeros
    if message.message_id == [0u8; 32] {
        return Err(ProgramError::InvalidInstructionData.into());
    }
    
    Ok(())
}

/// Extract NFT name from metadata URI (simplified implementation)
fn extract_name_from_uri(uri: &str) -> String {
    // This is a simplified implementation
    // In production, you might want to fetch the metadata and extract the name
    format!("Universal NFT")
}

/// Generate message ID for cross-chain compatibility
pub fn generate_message_id(
    sender: &[u8; 32],
    token_id: u64,
    destination_chain: u32,
    nonce: u64,
    timestamp: u64,
) -> [u8; 32] {
    let mut data = Vec::new();
    data.extend_from_slice(sender);
    data.extend_from_slice(&token_id.to_le_bytes());
    data.extend_from_slice(&destination_chain.to_le_bytes());
    data.extend_from_slice(&nonce.to_le_bytes());
    data.extend_from_slice(&timestamp.to_le_bytes());
    
    keccak::hash(&data).to_bytes()
}

/// Chain ID constants for cross-chain compatibility
pub mod chain_ids {
    pub const SOLANA_DEVNET: u32 = 900;
    pub const ZETACHAIN_TESTNET: u32 = 7001;
    pub const BASE_SEPOLIA: u32 = 84532;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_evm_format_conversion() {
        let message = NFTTransferMessage {
            token_id: 123,
            metadata_uri: "https://example.com/metadata.json".to_string(),
            recipient: [1u8; 32],
            origin_chain: chain_ids::ZETACHAIN_TESTNET,
            destination_chain: chain_ids::BASE_SEPOLIA,
            message_id: [2u8; 32],
            timestamp: 1640995200,
            origin_contract: [3u8; 32],
            nonce: 1,
        };

        let encoded = to_evm_format(&message);
        let decoded = from_evm_format(&encoded).unwrap();

        assert_eq!(message.token_id, decoded.token_id);
        assert_eq!(message.metadata_uri, decoded.metadata_uri);
        assert_eq!(message.recipient, decoded.recipient);
        assert_eq!(message.origin_chain, decoded.origin_chain);
        assert_eq!(message.destination_chain, decoded.destination_chain);
        assert_eq!(message.message_id, decoded.message_id);
        assert_eq!(message.timestamp, decoded.timestamp);
        assert_eq!(message.origin_contract, decoded.origin_contract);
        assert_eq!(message.nonce, decoded.nonce);
    }

    #[test]
    fn test_message_id_generation() {
        let sender = [4u8; 32];
        let token_id = 123u64;
        let destination_chain = chain_ids::BASE_SEPOLIA;
        let nonce = 1u64;
        let timestamp = 1640995200u64;

        let message_id1 = generate_message_id(&sender, token_id, destination_chain, nonce, timestamp);
        let message_id2 = generate_message_id(&sender, token_id, destination_chain, nonce, timestamp);
        
        assert_eq!(message_id1, message_id2);
        
        // Different parameters should generate different IDs
        let message_id3 = generate_message_id(&sender, token_id, destination_chain, nonce + 1, timestamp);
        assert_ne!(message_id1, message_id3);
    }

    #[test]
    fn test_validation() {
        let valid_message = NFTTransferMessage {
            token_id: 123,
            metadata_uri: "https://example.com/metadata.json".to_string(),
            recipient: [1u8; 32],
            origin_chain: chain_ids::ZETACHAIN_TESTNET,
            destination_chain: chain_ids::BASE_SEPOLIA,
            message_id: [2u8; 32],
            timestamp: Clock::get().unwrap().unix_timestamp as u64,
            origin_contract: [3u8; 32],
            nonce: 1,
        };

        assert!(validate_cross_chain_compatibility(&valid_message).is_ok());

        // Test invalid token ID
        let mut invalid_message = valid_message.clone();
        invalid_message.token_id = 0;
        assert!(validate_cross_chain_compatibility(&invalid_message).is_err());

        // Test same chain transfer
        let mut invalid_message = valid_message.clone();
        invalid_message.destination_chain = invalid_message.origin_chain;
        assert!(validate_cross_chain_compatibility(&invalid_message).is_err());
    }
}